src/QATools/Example/Pages/LoginPage.php

```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * This file is part of the QA-Tools library.
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ *
+ * @copyright Michael Iwersen <mi.iwersen@gmail.com>
+ * @link      https://github.com/qa-tools/behat-example
+ */
+
+namespace QATools\Example\Pages;
+
+
+use QATools\QATools\HtmlElements\Element\Form;
+use QATools\QATools\HtmlElements\TypifiedPage;
+
+/**
+ * Class LoginPage.
+ *
+ * @page-url('/')
+ */
+class LoginPage extends TypifiedPage
+{
+
+	/**
+	 * The login form.
+	 *
+	 * @var Form
+	 * @find-by('id' => 'login-form')
+	 */
+	protected $loginForm;
+
+	/**
+	 * Fill the login form with given data.
+	 *
+	 * @param string $username The username.
+	 * @param string $password The password.
+	 *
+	 * @return static
+	 */
+	public function fillForm($username, $password)
+	{
+		$this->loginForm->fill(array(
+			'username' => $username,
+			'password' => $password,
+		));
+
+		return $this;
+	}
+
+	/**
+	 * Does the login.
+	 *
+	 * @return Page
+	 */
+	public function loginSuccess()
```

Comment:
Also we're forgetting about login-required pages, where at the moment user tries to access protected page url he:

1. is redirected to login page with `?next=http://.....` in the url
2. on successful login he is redirected back to desired page

So, at least for login form the `PageObject` methods could be:

* `login($username, $password, $successPage = null, $failurePage = null)`
* `getLoginErrorMessage()`

Then scenario could look like:
```gherkin
Given I'm on 'Login protected page'
When I login with 'username' and 'password'
Then I should be on 'Login page'
And I should see 'invalid username or password' error message
```

Here is step mapping to PHP code in the FeatureContext class:

> Given I'm on "Login protected page"

```php
$this->page = $this->getPage('Login protected page');
$this->page->open();
```

> When I login with 'username' and 'password'

```php
$this->page = $this->page->login('username', 'password', null, 'Login Page');
```

> Then I should be on 'Login page'

```php
// 1. use PageLocator of PageFactory to transform 'Login Page' into actual class name
// 2. compare class of `$this->page` with resolved class name from step 1
$this->page->matches('Login Page');

// 3. do some fancy url matching of actually opened page with url of `$this->page`
$this->page->opened();
```

> And I should see 'invalid username or password' error message

```php
$this->assertEquals($this->page->getLoginErrorMessage(), 'invalid username or password');
```